proposals/22-04-localize-command.md

```diff
@@ -0,0 +1,335 @@
+functions. Downloaded images would live only in local caches and kustomize cannot know the form that custom fields will
+take. Thus, neither is worth localizing.
+
+**Goals:**
+
+1. This command should localize 
+   * all remote files that a kustomization file directly references 
+   * remote exec binaries of referenced KRM functions 
+
+   This command achieves this goal if, in the absence of remote images and custom fields in KRM
+   functions, `kustomize build` can run on the localized copy without network access.
+
+**Non-goals:**
+
+1. This command should not localize remote images or custom fields in KRM functions.
+2. This command should not copy files that the target kustomization does not reference.
+3. This command should not serve as a package manager.
+
+## Proposal
+
+The command takes the following form:
+
+<pre>
+<b>kustomize localize</b> <ins>target</ins> <ins>scope</ins> <ins>newDir</ins>
+</pre>
+
+where the arguments are:
+
+* `target`: a directory with a top-level kustomization file that kustomize will localize; can be a path to a local
+  directory or a url to a remote directory
+* `scope`: optional root directory, files outside which kustomize is not allowed to copy and localize; if not specified,
+  takes on value of `target`
+* `newDir`: destination directory of the localized copy of `target`
+
+The command creates a copy of the `target` kustomization and the local files that `target` references at `newDir`. We
+define the "files that `target` references" as:
+
+* kustomization files that `target` directly or transitively references
+* configuration files that referenced kustomization files reference
+* exec binaries of referenced KRM functions
+
+Here, configuration file means a non-kustomization yaml file. The command only copies referenced files that reside
+inside `scope`.
+
+The command localizes the copy of `target` at `newDir` by downloading all remote files that `target` references. The
+downloaded files placed in a new `localized-files` directory next to the file that referenced the downloaded files.
+Inside `localized-files`, the downloads are located on path:
+
+<pre>
+<ins>remote-host</ins> / <ins>organization</ins> / <ins>repo</ins> / <ins>version</ins> / <ins>path-to-file-in-repo</ins>
+</pre>
+
+The command rewrites remote references in `newDir` to the local paths of the downloaded files. To help ensure
+that `newDir` is a clean copy, the command additionally overwrites absolute path references into `target` to point
+to `newDir`.
+
+**Error cases**:
+* `target` does not have a top-level kustomization file
+* `scope` does not contain `target`
+* `newDir` already exists
+* `localized-files` directory already exists
+* remote reference does not have a version
+* kustomization file is malformed
+* cycle of kustomization file references exists
+
+**Warning cases**:
+* `target` references a local path that traverses outside of `scope`
```

Comment:
I leaned towards a warning for uses cases like **User Story 1**, where the user wants `newDir` to be next to `target` so that the two can share kustomizations and configurations higher up in the directory tree. I wasn't sure if certain use cases may not want to copy the entire tree. I'm also curious if certain use cases may not have access to the entire tree.

Reply:
Let's go with an error. As a rule, it's easier to start more strict and relax later if needed, than the other way around.